### PR TITLE
[CSM] Add ModStorageAPI 

### PR DESCRIPTION
**Pull request by nerzhul**
Related to #5394

### CODE REVIEW

**Comment by nerzhul** [comment]
It's now ready for review & integration, @sofar @Zeno-

File: src/unittest/test.cpp [hunk 1]
```diff
@@ -65,6 +65,9 @@ class TestGameDef : public IGameDef {
 		return testmodspec;
 	}
 	virtual const ModSpec* getModSpec(const std::string &modname) const { return NULL; }
+	virtual std::string getModStoragePath() const { return "/tmp"; }
```
Comment (SmallJoker) [inline src/unittest/test.cpp]: Not important at the moment but may cause problems for Windows users when running the unittests (once they test the mod storage).

**Comment by nerzhul** [comment]
Yeah maybe i should set "." Instead

File: src/unittest/test.cpp (hunk 1 above)
Comment (nerzhul) [inline src/unittest/test.cpp]: I changed it to "." it works well, tested on my machine

**Comment by nerzhul** [comment]
It's now okay for merge.
Preview mod shows it works like in server, (we store modname and retrieve it)

File: src/client.cpp [hunk 2]
```diff
@@ -1998,3 +2011,31 @@ scene::IAnimatedMesh* Client::getMesh(const std::string &filename)
...
+	if (it != m_mod_storages.end()) {
+		// Save unconditionaly on unregistration
+		it->second->save(getModStoragePath());
+		m_mod_storages.erase(name);
+	}
+}
+
+std::string Client::getModStoragePath() const
+{
+	return porting::path_user + DIR_DELIM  "client" + DIR_DELIM + "mod_storage";
```
Comment (bigfoot547) [inline src/client.cpp]: I am new to c++ but should there be a `+` between `DIR_DELIM` and `"client"`

Comment (nerzhul) [inline src/client.cpp]: nice shot, compiler accept it because it's static variables, but for coherence it's needed

File: src/client.cpp [hunk 3]
```diff
@@ -730,6 +731,18 @@ void Client::step(float dtime)
 		}
 	}
 
+	m_mod_storage_save_timer -= dtime;
+	if (m_mod_storage_save_timer <= 0.0f) {
+		infostream << "Saving registered mod storages." << std::endl;
+		m_mod_storage_save_timer = g_settings->getFloat("server_map_save_interval");
+		for (UNORDERED_MAP<std::string, ModMetadata *>::const_iterator
+			it = m_mod_storages.begin(); it != m_mod_storages.end(); ++it) {
```
Comment (SmallJoker) [inline src/client.cpp]: One indent missing

**Comment by nerzhul** [comment]
Fixed, we can now merge @SmallJoker

**Comment by paramat** [comment]
A description of what this does would help reviewers.

**Comment by nerzhul** [comment]
@paramat it's the same API than recently merged server side API: you get your own mod container, write your data and read it

File: src/client.cpp [hunk 4]
```diff
@@ -730,6 +731,18 @@ void Client::step(float dtime)
 		}
 	}
 
+	m_mod_storage_save_timer -= dtime;
+	if (m_mod_storage_save_timer <= 0.0f) {
+		infostream << "Saving registered mod storages." << std::endl;
```
Comment (sofar) [inline src/client.cpp]: debugstream? verbosestream?

File: src/client.cpp [hunk 5]
```diff
@@ -1998,3 +2011,31 @@ scene::IAnimatedMesh* Client::getMesh(const std::string &filename)
 	smgr->getMeshCache()->removeMesh(mesh);
 	return mesh;
 }
+
+bool Client::registerModStorage(ModMetadata *storage)
+{
+	if (m_mod_storages.find(storage->getModName()) != m_mod_storages.end()) {
+		errorstream << "Unable to register same mod storage twice. Storage name: "
+				<< storage->getModName() << std::endl;
+		return false;
```
Comment (sofar) [inline src/client.cpp]: hmmm, shouldn't we make this fail instead? do we really want mods to sort this problem out?

File: src/client.cpp (hunk 4 above)
Comment (nerzhul) [inline src/client.cpp]: moving to verbosestream okay :)

File: src/client.cpp (hunk 5 above)
Comment (nerzhul) [inline src/client.cpp]: failing is not a good idea, especially on android devices where user cannot do anything

**Pull request merged**